libs/core/libs/shared/src/document_repo.rs

```diff
@@ -1,15 +1,156 @@
 use crate::core_config::Config;
-use crate::crypto::*;
 use crate::file_metadata::DocumentHmac;
-use crate::{SharedErrorKind, SharedResult};
-use std::collections::HashSet;
+use crate::SharedResult;
+use crate::{crypto::*, SharedErrorKind};
+use serde::{Deserialize, Serialize};
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
 use std::convert::TryInto;
 use std::fs::{self, File, OpenOptions};
+use std::hash::Hash;
 use std::io::{ErrorKind, Read, Write};
 use std::path::Path;
 use tracing::*;
 use uuid::Uuid;
 
+#[derive(Debug, Clone, Serialize, Deserialize, Ord, PartialEq, PartialOrd, Eq, Hash)]
+pub enum DocEvent {
+    Read(Uuid, i64),
+    Write(Uuid, i64),
+}
+impl DocEvent {
+    pub fn timestamp(&self) -> i64 {
+        match *self {
+            DocEvent::Read(_, timestamp) => timestamp,
+            DocEvent::Write(_, timestamp) => timestamp,
+        }
+    }
+    pub fn id(&self) -> Uuid {
+        match *self {
+            DocEvent::Read(id, _) => id,
+            DocEvent::Write(id, _) => id,
+        }
+    }
+}
+
+#[derive(Default, Copy, Clone, PartialEq)]
+pub struct StatisticValue {
+    pub raw: i64,
+    pub normalized: f64,
+}
+impl StatisticValue {
+    fn from_raw(raw: i64) -> StatisticValue {
+        StatisticValue { raw, normalized: f64::default() }
```

Comment:
I would personally either
1. not have this exist until you can calculate the normalized value
2. make `normalized: Option<f64>` to represent it's un-calculated state

usually I try to do 1